src/session/session_dhandle.c

```diff
@@ -357,6 +363,8 @@ __wt_session_get_btree(WT_SESSION_IMPL *session,
 	bucket = __wt_hash_city64(uri, strlen(uri)) % WT_HASH_ARRAY_SIZE;
 	SLIST_FOREACH(dhandle_cache, &session->dhhash[bucket], hashl) {
 		dhandle = dhandle_cache->dhandle;
+		if (F_ISSET(dhandle, WT_DHANDLE_DEAD))
+			continue;
```

Comment:
I understand what you're saying -- at the moment we combine freeing the `WT_DATA_HANDLE_CACHE` structure and decrementing the `session_ref` for handles.  If we split that, we could swap in the new handle immediately here.  I'll take a look...